### PR TITLE
WIP: Address Crystal 0.36.1 complaints and latest alea update

### PR DESCRIPTION
**Pull request by lbarasti**
Hi folks, this is to start the conversation on what needs to be done to ensure `num.cr` works with the latest version of `alea` (which [now requires Crystal v0.36.0 or greater](https://github.com/nin93/alea/commits/master)).

See https://github.com/crystal-data/num.cr/issues/63

### CODE REVIEW

File: src/grad/primitives/gate.cr [hunk 1]
```diff
@@ -34,5 +34,5 @@ abstract class Num::Grad::Gate(T)
   abstract def backward(payload : Num::Grad::Payload(T)) : Array(T)
 
   # Caches the result of an operation on a context
-  abstract def cache(result : Num::Grad::Variable(T), *args)
+  abstract def cache(result : Num::Grad::Variable(T), *args : Num::Grad::Variable(T))
```
Comment (christopherzimmerman) [inline src/grad/primitives/gate.cr:37]: Is the issue with this that child methods sometimes say `args` is an instance of `Num::Grad::Variable`, which isn't explicitly defined by the abstract base, or that `args` must now have a defined type?

`cache` methods can take anything that a gate needs to store information.  For example, the convolutional gate also stores tuples of integers which match up with padding and stride.

**Comment by christopherzimmerman** [comment]
Thank you for the PR.  I haven't had time to work much on this library recently, but I will double check tomorrow to make sure everything works with the most recent version.

File: src/grad/primitives/gate.cr (hunk 1 above)
Comment (lbarasti) [reply src/grad/primitives/gate.cr:37]: Hey Chris, the issue is the compiler now complains that the implementations are too strict
```
In src/grad/gates_blas.cr:25:1

 25 | class Num::Grad::MatMulGate(T) < Num::Grad::Gate(T)
      ^
Error: abstract `def Num::Grad::Gate(T)#cache(result : Num::Grad::Variable(T), *args)` must be implemented by Num::Grad::MatMulGate(T)
```

So either we restrict the abstract def or we remove the type restriction on each `def cache` definition

image

**Comment by christopherzimmerman** [comment]
@lbarasti closing this in favor of the now merged #65.  I still need to do some work with library dependencies, to make transitions smoother.

**Pull request closed without merging**